renku/cli/_checks/migrate_datasets.py

```diff
@@ -0,0 +1,226 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Checks needed to determine dataset migration policy."""
+import shutil
+import uuid
+from collections import defaultdict
+from pathlib import Path
+
+import click
+
+from renku.models.datasets import Dataset
+from renku.models.refs import LinkReference
+
+from .._echo import WARNING
+
+
+def dataset_pre_0_3(client):
+    """Return paths of dataset metadata for pre 0.3.4."""
+    return (client.path / 'data').rglob(client.METADATA)
+
+
+def check_dataset_metadata(client):
+    """Check location of dataset metadata."""
+    # Find pre 0.3.4 metadata files.
+    old_metadata = list(dataset_pre_0_3(client))
+
+    if not old_metadata:
+        return True
+
+    click.secho(
+        WARNING + 'There are metadata files in the old location.'
+        '\n  (use "renku migrate datasets" to move them)\n\n\t' + '\n\t'.join(
+            click.style(str(path.relative_to(client.path)), fg='yellow')
+            for path in old_metadata
+        ) + '\n'
+    )
+
+    return False
+
+
+def check_missing_files(client):
+    """Find missing files listed in datasets."""
+    missing = defaultdict(list)
+
+    for path, dataset in client.datasets.items():
+        for file_ in dataset.files:
+            filepath = Path(file_.path)
+            if not filepath.exists():
+                missing[str(
+                    path.parent.relative_to(client.renku_datasets_path)
+                )].append(str(filepath))
+
+    if not missing:
+        return True
+
+    click.secho(
+        WARNING + 'There are missing files in datasets.'
+        # '\n  (use "renku dataset clean <name>" to clean them)'
+    )
+
+    for dataset, files in missing.items():
+        click.secho(
+            '\n\t' + click.style(dataset, fg='yellow') + ':\n\t  ' +
+            '\n\t  '.join(click.style(path, fg='red') for path in files)
+        )
+
+    return False
+
+
+def check_dataset_resources(client):
+    """Find missing datasets or files listed in metadata."""
+    missing_datasets = defaultdict(list)
+    missing_files = defaultdict(list)
+
+    for path, dataset in client.datasets.items():
+
+        if not (Path(dataset.path) / Path(client.METADATA)).exists():
+            missing_datasets[path] = dataset
+
+        for file_ in dataset.files:
+            filepath = Path(file_.path)
+            if not filepath.exists():
+                relative = path.parent.relative_to(client.renku_datasets_path)
+                missing_files[str(relative)].append(str(filepath))
+
+    return missing_datasets, missing_files
+
+
+def ensure_clean_lock(client):
+    """Make sure Renku lock file is not part of repository."""
+    lock_file = Path('.renku.lock')
+    if lock_file.exists():
```

Comment:
This should check if the lock file is in the git tree, not just on the filesystem. If it's been checked in and deleted then the current code doesn't remove it from git. If you use my suggestion below then the code fails when the file is present on the filesystem but it's not checked in git. A better approach might be:

```
    lock_file = Path('.renku.lock')
    try:
        client.repo.tree() / '.renku.lock'
    except KeyError:  # the lock file is not checked in
        return
    lock_file.unlink()
    client.repo.git.add(str(lock_file))
```

Reply:
Well since you cannot run the structural migration without a clean repository, this check will actually check if its part of the git tree, since you will always run it from the HEAD clean state.

Reply:
It isn't because it doesn't actually remove the file from git if it's checked in.

Reply:
Hmm, I'm failing to reproduce what you're reporting. I've added the test which validates what I said [above](https://github.com/SwissDataScienceCenter/renku-python/pull/639/files#diff-7e3313f0ee1c3b9530e15596b9b67537R122). Also I've added to that migration that we ensure that lock file is part of the gitignore.

Reply:
you're right, with the current changes it seems to work 🤞